### ISSUE TITLE
The `done` part of an `EventSource` is run more often than expected

### ISSUE THREAD

**Issue by menelaos** [slamdata/purescript-halogen#418]
I was trying to figure out Halogen's `EventSource` mechanism and ran into an issue which might be a bug (I'm not sure).

TL;DR: The `done` part of an `EventSource` is run more often than I would expect.

To illustrate the issue I've created a repo [here](http://www.github.com/menelaos/halogen-issue). Just run `pulp browserify > app.js` and e.g. `python -m http.server`.

Basically, I have a `Parent` component that can switch between two child views. To keep it simple, `ChildB` only displays text. `ChildA` is a `lifecycleComponent` which subscribes to a `busEventSource` upon initialization. The bus is part of the reader environment (`Wiring`) and emits `Unit` values. Whenever the bus fires, a `Toggle` query is raised in `ChildA` which toggles its `On/Off` state. After 10 seconds, the bus receives a `Bus.write` in the `main` function. The `busEventSource` function is taken from SlamData's `Halogen.Component.Utils` module. I've modified the `done` function in `affEventSource` in order to log a random number to the console whenever it is run.

Here's how to reproduce the issue:
* Open the console
* Load the page
* Within the first 10 seconds, switch between `ChildA` and `ChildB` a few times.

You will notice that each time `ChildA` is un-mounted, the `done` function is run as expected. After 10 seconds, however, all `done` functions are run again. I.e. if `ChildA` has been mounted/unmounted `n` times, `n` `done` functions will be run after 10 seconds in addition to those that have been run immediately upon unmounting.

Note that this does not affect the `Toggle` query in `ChildA` which is correctly dispatched only once after 10 seconds (if the component is mounted, of course).

Is this behaviour intended?
Thanks for your time!

**Comment by garyb** [comment]
It's not intentional, no... thanks for putting the reproduction case together, I'll take a look.

In some ways it's not too bad at least, as the `done` function is meant as a clean-up thing to unhook event listeners, etc. so should be idempotent. It would be definitely better to avoid calling it again though!

**Comment by natefaubion** [comment]
This is probably because of how we use `forkAff` to spawn the subscription (and also a bit how `busEventSource` works). https://github.com/slamdata/purescript-halogen/blob/master/src/Halogen/Aff/Driver/Eval.purs#L115

The forked thread contains the full lifecycle. When you unmount the component, the thread isn't cleaned up, so it's still `await`ing. You push to the bus, which fulfills the `await`, the component is no longer active, and so it calls `done`. One option may be to `kill` all pending subscription threads. `EventLoop` could also probably use some work, since when you `break` it, it doesn't actually kill the loop until it comes back around to the head.
